src/Compilers/Core/Portable/CodeGen/DataStringHolder.cs

```diff
@@ -0,0 +1,424 @@
+    }
+
+    private Cci.IMethodDefinition GetOrSynthesizeBytesToStringHelper(DiagnosticBag diagnostics)
+    {
+        var method = _privateImplementationDetails.GetMethod(PrivateImplementationDetails.SynthesizedBytesToStringFunctionName);
+
+        if (method is null)
+        {
+            _privateImplementationDetails.TryAddSynthesizedMethod(SynthesizeBytesToStringHelper(diagnostics));
+
+            method = _privateImplementationDetails.GetMethod(PrivateImplementationDetails.SynthesizedBytesToStringFunctionName);
+        }
+
+        Debug.Assert(method is not null);
+        return method;
+    }
+
+    public Cci.IUnitReference GetUnit(EmitContext context)
+    {
+        Debug.Assert(context.Module == _moduleBuilder);
+        return _moduleBuilder;
+    }
+
+    private BytesToStringHelper SynthesizeBytesToStringHelper(DiagnosticBag diagnostics)
+    {
+        var ilBuilder = new ILBuilder((ITokenDeferral)_moduleBuilder, new LocalSlotManager(slotAllocator: null), OptimizationLevel.Release, areLocalsZeroed: false);
+
+        // Call `Encoding.get_UTF8()`.
+        ilBuilder.EmitOpCode(ILOpCode.Call, 1);
+        ilBuilder.EmitToken(_encodingUtf8, null, diagnostics);
+
+        // Push the `byte*`.
+        ilBuilder.EmitOpCode(ILOpCode.Ldarg_0);
+
+        // Push the byte size.
+        ilBuilder.EmitOpCode(ILOpCode.Ldarg_1);
+
+        // Call `Encoding.GetString(byte*, int)`.
+        ilBuilder.EmitOpCode(ILOpCode.Callvirt, -2);
+        ilBuilder.EmitToken(_encodingGetString, null, diagnostics);
+
+        // Return.
+        ilBuilder.EmitRet(isVoid: false);
+        ilBuilder.Realize();
+
+        return new BytesToStringHelper(
+            containingType: _privateImplementationDetails,
+            encodingGetString: _encodingGetString,
+            maxStack: ilBuilder.MaxStack,
+            il: ilBuilder.RealizedIL);
+    }
+
+    private StaticConstructor SynthesizeStaticConstructor(DiagnosticBag diagnostics)
+    {
+        var ilBuilder = new ILBuilder((ITokenDeferral)_moduleBuilder, new LocalSlotManager(slotAllocator: null), OptimizationLevel.Release, areLocalsZeroed: false);
+
+        Cci.IMethodDefinition? bytesToStringHelper = null;
+
+        foreach (var field in _orderedSynthesizedFields)
+        {
+            if (field is not StringField stringField)
+            {
+                continue;
+            }
+
+            // Push the `byte*` field's address.
+            ilBuilder.EmitOpCode(ILOpCode.Ldsflda);
+            ilBuilder.EmitToken(stringField.MappedField, null, diagnostics);
+
+            // Push the byte size.
+            ilBuilder.EmitIntConstant(stringField.MappedField.MappedData.Length);
+
+            // Call `<PrivateImplementationDetails>.BytesToString(byte*, int)`.
+            bytesToStringHelper ??= GetOrSynthesizeBytesToStringHelper(diagnostics);
+            ilBuilder.EmitOpCode(ILOpCode.Call, -1);
+            ilBuilder.EmitToken(bytesToStringHelper, null, diagnostics);
```

Comment:
>bytesToStringHelper [](http://example.com/codeflow?start=32&length=19)

Given that we only ever have one field to initialize, it doesn't feel like we need this helper method. We can simply inline what it is doing